src/command/archive/find.c

```diff
@@ -0,0 +1,230 @@
+    StringList *list;                                               // List of found segments
+};
+
+/***********************************************************************************************************************************
+Macros for function logging
+***********************************************************************************************************************************/
+#define FUNCTION_LOG_WAL_SEGMENT_FIND_TYPE                                                                                         \
+    WalSegmentFind *
+#define FUNCTION_LOG_WAL_SEGMENT_FIND_FORMAT(value, buffer, bufferSize)                                                            \
+    objNameToLog(value, "WalSegmentFind", buffer, bufferSize)
+
+/**********************************************************************************************************************************/
+FN_EXTERN WalSegmentFind *
+walSegmentFindNew(const Storage *const storage, const String *const archiveId, const bool single, const TimeMSec timeout)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(STORAGE, storage);
+        FUNCTION_LOG_PARAM(STRING, archiveId);
+        FUNCTION_LOG_PARAM(BOOL, single);
+        FUNCTION_LOG_PARAM(TIME_MSEC, timeout);
+    FUNCTION_LOG_END();
+
+    ASSERT(storage != NULL);
+    ASSERT(archiveId != NULL);
+
+    OBJ_NEW_BEGIN(WalSegmentFind, .childQty = MEM_CONTEXT_QTY_MAX)
+    {
+        *this = (WalSegmentFind)
+        {
+            .storage = storage,
+            .archiveId = strDup(archiveId),
+            .single = single,
+            .timeout = timeout,
+        };
+    }
+    OBJ_NEW_END();
+
+    FUNCTION_LOG_RETURN(WAL_SEGMENT_FIND, this);
+}
+
+/**********************************************************************************************************************************/
+FN_EXTERN String *
+walSegmentFind(WalSegmentFind *const this, const String *const walSegment)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(WAL_SEGMENT_FIND, this);
+        FUNCTION_LOG_PARAM(STRING, walSegment);
+    FUNCTION_LOG_END();
+
+    ASSERT(walSegment != NULL);
+    ASSERT(walIsSegment(walSegment));
+
+    String *result = NULL;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        Wait *const wait = waitNew(this->timeout);
+        const String *const prefix = strSubN(walSegment, 0, 16);
+        const String *const path = strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveId), strZ(prefix));
+        const String *const expression = strNewFmt(
+            "^%s%s-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$", strZ(strSubN(walSegment, 0, 24)),
+            walIsPartial(walSegment) ? WAL_SEGMENT_PARTIAL_EXT : "");
+        RegExp *regExp = NULL;
+
+        do
+        {
+            // Get a list of all WAL segments that match the directory (and prefix when finding a single WAL)
+            if (this->list == NULL || !strEq(prefix, this->prefix))
```

Comment:
Given that this is testing for this->list being NULL and not merely empty, isn't it the case that eventually the list will be empty due to the removals and then a later search for a segment (for whatever reason, if it happens with this WalSegmentFind...) is attempted, it won't be found even if it actually existed?

More broadly, I'd say this function should probably have a more specific name if it's going to be "grab the list of segments with this prefix and then go through them but only allow any to be checked once" or if it's intended as "general walSegmentFind but with an optimization to grab a whole list and then go through it in order, but will still work properly for any requested segment".

Also just generally, not sure if the delete is really necessary anyway?  One could imagine that you just remember the last position and step forward from there when searching, which would allow you to wrap around if an earlier segment was asked for again for some reason, and also might be cheaper than the routine delete's and associated memmove's.

Reply:
> Given that this is testing for this->list being NULL and not merely empty, isn't it the case that eventually the list will be empty due to the removals and then a later search for a segment (for whatever reason, if it happens with this WalSegmentFind...) is attempted, it won't be found even if it actually existed?

The list is cleared and set to NULL at line 178 when there is nothing left in it, but there was no explicit test for this so added at 5cc0030a50.

> More broadly, I'd say this function should probably have a more specific name

As far as I can see, the functionality has not changed, except that it is possible to optimize it for a single lookup or multiple lookups. Single lookups still work in the general case, they just have to check more results. Conversely, multiple lookups still work if single lookup is specified, but they will be no faster than the prior code.

Also, when doing multiple lookups the optimization only works well if segments are requested in order, but in any case the results should be correct.

> Also just generally, not sure if the delete is really necessary anyway?

In the List object, deleting the first item just moves the pointer (https://github.com/pgbackrest/pgbackrest/blob/release/2.47/src/common/type/list.c#L339) so no reason for us to reinvent that in this function.